> Can you add some unit test cases to cover these three special checks? I'm thinking it should be like this test case, just change `version`:

Thanks for the pointer, it made it very easy.  I've added the tests.  I didn't add 3 new plugins in the plugin/testdata directory as the test still fail appropriately as follows, before the fix from this PR:
```
--- FAIL: TestKubectlCommandHandlesPlugins (0.08s)
    --- FAIL: TestKubectlCommandHandlesPlugins/test_that_a_plugin_does_not_execute_over_Cobra's_help_command (0.01s)
        cmd_test.go:113: unexpected error: expected "" to occur, but got "unable to find a plugin executable \"kubectl-help\""
    --- FAIL: TestKubectlCommandHandlesPlugins/test_that_a_plugin_does_not_execute_over_Cobra's___complete_command (0.01s)
        cmd_test.go:113: unexpected error: expected "" to occur, but got "unable to find a plugin executable \"kubectl-__complete\""
    --- FAIL: TestKubectlCommandHandlesPlugins/test_that_a_plugin_does_not_execute_over_Cobra's___completeNoDesc_command (0.01s)
        cmd_test.go:113: unexpected error: expected "" to occur, but got "unable to find a plugin executable \"kubectl-__completeNoDesc\""
```

> As for backward compatibility, I can't think of a legitimate reason why we would want to allow a plugin named `kubectl-__complete` (or the other two) to run.

👍

--- a/staging/src/k8s.io/kubectl/pkg/cmd/cmd.go
+++ b/staging/src/k8s.io/kubectl/pkg/cmd/cmd.go
@@ -105,7 +105,15 @@ func NewDefaultKubectlCommandWithArgs(pluginHandler PluginHandler, args []string
 			// Also check the commands that will be added by Cobra.
 			// These commands are only added once rootCmd.Execute() is called, so we
 			// need to check them explicitly here.
-			switch cmdPathPieces[0] {
+			var cmdName string // first "non-flag" arguments
+			for _, arg := range cmdPathPieces {
+				if !strings.HasPrefix(arg, "-") {
+					cmdName = arg
+					break
+				}
+			}
+
+			switch cmdName {
 			case "help", cobra.ShellCompRequestCmd, cobra.ShellCompNoDescRequestCmd:
 				// Don't search for a plugin
 			default:

--- a/staging/src/k8s.io/kubectl/pkg/cmd/cmd.go
+++ b/staging/src/k8s.io/kubectl/pkg/cmd/cmd.go
@@ -102,9 +102,17 @@ func NewDefaultKubectlCommandWithArgs(pluginHandler PluginHandler, args []string
 		// only look for suitable extension executables if
 		// the specified command does not already exist
 		if _, _, err := cmd.Find(cmdPathPieces); err != nil {
-			if err := HandlePluginCommand(pluginHandler, cmdPathPieces); err != nil {
-				fmt.Fprintf(errout, "Error: %v\n", err)
-				os.Exit(1)
+			// Also check the commands that will be added by Cobra.
+			// These commands are only added once rootCmd.Execute() is called, so we
+			// need to check them explicitly here.
+			switch cmdPathPieces[0] {
+			case "help", cobra.ShellCompRequestCmd, cobra.ShellCompNoDescRequestCmd:
+				// Don't search for a plugin
+			default:
+				if err := HandlePluginCommand(pluginHandler, cmdPathPieces); err != nil {
+					fmt.Fprintf(errout, "Error: %v\n", err)
+					os.Exit(1)
+				}
 			}
 		}
 	}

--- a/staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go
+++ b/staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go
@@ -79,6 +79,7 @@ func TestKubectlCommandHandlesPlugins(t *testing.T) {
 			args: []string{"kubectl", "version"},
 		},
 		// The following tests make sure that commands added by Cobra cannot be shadowed by a plugin
+		// See https://github.com/kubernetes/kubectl/issues/1116
 		{
 			name: "test that a plugin does not execute over Cobra's help command",
 			args: []string{"kubectl", "help"},
@@ -91,6 +92,36 @@ func TestKubectlCommandHandlesPlugins(t *testing.T) {
 			name: "test that a plugin does not execute over Cobra's __completeNoDesc command",
 			args: []string{"kubectl", cobra.ShellCompNoDescRequestCmd},
 		},
+		// The following tests make sure that commands added by Cobra cannot be shadowed by a plugin
+		// even when a flag is specified first.  This can happen when using aliases.
+		// See https://github.com/kubernetes/kubectl/issues/1119
+		{
+			name: "test that a flag does not break Cobra's help command",
+			args: []string{"kubectl", "--kubeconfig=/path/to/kubeconfig", "help"},
+		},
+		{
+			name: "test that a flag does not break Cobra's __complete command",
+			args: []string{"kubectl", "--kubeconfig=/path/to/kubeconfig", cobra.ShellCompRequestCmd},
+		},
+		{
+			name: "test that a flag does not break Cobra's __completeNoDesc command",
+			args: []string{"kubectl", "--kubeconfig=/path/to/kubeconfig", cobra.ShellCompNoDescRequestCmd},
+		},
+		// As for the previous tests, an alias could add a flag without using the = form.
+		// We don't support this case as parsing the flags becomes quite complicated (flags
+		// that take a value, versus flags that don't)
+		// {
+		// 	name: "test that a flag with a space does not break Cobra's help command",
+		// 	args: []string{"kubectl", "--kubeconfig", "/path/to/kubeconfig", "help"},
+		// },
+		// {
+		// 	name: "test that a flag with a space does not break Cobra's __complete command",
+		// 	args: []string{"kubectl", "--kubeconfig", "/path/to/kubeconfig", cobra.ShellCompRequestCmd},
+		// },
+		// {
+		// 	name: "test that a flag with a space does not break Cobra's __completeNoDesc command",
+		// 	args: []string{"kubectl", "--kubeconfig", "/path/to/kubeconfig", cobra.ShellCompNoDescRequestCmd},
+		// },
 	}
 
 	for _, test := range tests {
@@ -105,6 +136,7 @@ func TestKubectlCommandHandlesPlugins(t *testing.T) {
 			})
 
 			root := NewDefaultKubectlCommandWithArgs(pluginsHandler, test.args, in, out, errOut)
+			root.SetOut(out)
 			if err := root.Execute(); err != nil {
 				t.Fatalf("unexpected error: %v", err)
 			}

--- a/staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go
+++ b/staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go
@@ -78,6 +78,19 @@ func TestKubectlCommandHandlesPlugins(t *testing.T) {
 			name: "test that a plugin does not execute over an existing command by the same name",
 			args: []string{"kubectl", "version"},
 		},
+		// The following tests make sure that commands added by Cobra cannot be shadowed by a plugin
+		{
+			name: "test that a plugin does not execute over Cobra's help command",
+			args: []string{"kubectl", "help"},
+		},
+		{
+			name: "test that a plugin does not execute over Cobra's __complete command",
+			args: []string{"kubectl", cobra.ShellCompRequestCmd},
+		},
+		{
+			name: "test that a plugin does not execute over Cobra's __completeNoDesc command",
+			args: []string{"kubectl", cobra.ShellCompNoDescRequestCmd},
+		},
 	}
 
 	for _, test := range tests {

